src/throttler/throttler.ts

```diff
@@ -254,4 +232,73 @@ export abstract class Throttler<T, R> {
       return p.resolve();
     });
   }
+
+  private initializeTimeout(taskData: TaskData<T, R>, cursorIndex: number): Promise<void> {
+    const timeoutMillis = taskData.timeoutMillis!;
+    const timeoutPromise = new Promise<void>((_, reject) => {
+      taskData.timeoutId = setTimeout(() => {
+        taskData.isTimedOut = true;
+        reject(new TimeoutError(this.taskName(cursorIndex), timeoutMillis));
+      }, timeoutMillis);
+    });
+
+    return timeoutPromise;
+  }
+
+  private async executeTask(taskData: TaskData<T, R>, cursorIndex: number): Promise<any> {
+    const t0 = Date.now();
+    try {
+      const result = await this.handler(taskData.task);
+      const dt = Date.now() - t0;
+      this.min = Math.min(dt, this.min);
+      this.max = Math.max(dt, this.max);
+      this.avg = (this.avg * this.complete + dt) / (this.complete + 1);
+
+      this.success++;
+      this.complete++;
+      this.active--;
```

Comment:
Can we touch these counters only in one method, preferably `handle`.

Right now, I can find cases that these counters get messed due to racing conditions:
1. Timed out, so the `this.fail++`
2. Then the task succeeds, so `this.success++`

Let's do the counters after `Promise.race()` so only one of them get executed.

Reply:
I moved the counters to `handle`